hack/update-helm.sh

```diff
@@ -157,6 +157,7 @@ for output_file in ${DEST_CRD_DIR}/*.yaml; do
     fi
   done <"$input_file"
   rm $input_file
+  $SED -i '/^metadata:.*/a\  labels:\n  {{- include "kueue.labels" . | nindent 4 }}' $output_file
```

Comment:
The sed manipulations are tricky. Can this corrupt the output file if we already have some metadata.labels? 
I'm thinking if there are other tools which understand yaml, like YQ which could be employed to do this, but not sure, can you make a short summary of alternatives?

Reply:
I agree. I tried first to use YQ, but could not find a solution. I think the fundamental issue is that Helm templates are not valid templates, so in that particular case - adding a Helm template entry in a map - I've failed to implement it using JQ.

An alternative might be to revisit how the Kustomize manifests -> Helm charts script works, by taking the "rendered" Kustomize resources as input instead of the "raw" files themselves.

Reply:
> I agree. I tried first to use YQ, but could not find a solution. I think the fundamental issue is that Helm templates are not valid templates, so in that particular case - adding a Helm template entry in a map - I've failed to implement it using JQ.

I see, but this is interesting why we can use YQ a couple of lines above in the same script, using `$YQ -N -i`. Is it because the yaml files get invalid yaml's while being processed?

> An alternative might be to revisit how the Kustomize manifests -> Helm charts script works, by taking the "rendered" Kustomize resources as input instead of the "raw" files themselves.

Not sure I understand, IIUC if we used rendered files, then wouldn't we hit the same issue, that some replacements corrupt the file, and yq cannot be used more?

Reply:
> I see, but this is interesting why we can use YQ a couple of lines above in the same script, using $YQ -N -i. Is it because the yaml files get invalid yaml's while being processed?

You're right, the existing `$YQ -N -I` set single field values and that works fine. But for inlining labels, it touches the structure of the YAML document itself. It possibly is my fairly limited knowledge about YQ, but an explanation why that might not be possible is that for the former it's valid YAML being processed, e.g.:

```
metadata:
  namespace: "{{ .Release.Namespace }}"
```

While for the latter, it's not:

```
labels:
{{- include "kueue.labels" . | nindent 4 }}
```

Reply:
> Not sure I understand the idea, IIUC if we used rendered files, then wouldn't we hit the same issue, that some replacements corrupt the file, and yq cannot be used more?

You're right, I realise my idea would be more the other way, i.e., rendered Helm Charts  -> Kustomize manifests, so it's always valid YAML that's processed. But using Helm manifests as the source of truth is probably very debatable :).

Reply:
This makes sense, but I'm also not sure how to do it better, let's see if @tenzen-y or @alculquicondor have some ideas.

Reply:
TBH, I'm hesitating this way (rendered Helm Charts -> Customize manifests) since IIUC, the controller-tools, don't support automatic generation of manifests. Also, I suspect that using Helm Chart (programmable YAML) as the source of truth would increase maintenance costs.

As an alternative way, we might be possible to implement a simple tool to generate the Helm Charts from kustomize in Go or Python instead of this shell script.

Reply:
IIUC, then in the script we could load the entire valid yaml, do all replacements and output the helm chart, which is no longer valid yaml, but we don't care at this point, because we already did all processing.

Reply:
TBH, this script is quite unreadable already.
It would be nice to migrate it to golang, but for the time being, I'm ok with this.

Reply:
I agree with you.

Reply:
I'm ok with this for now as well, but I think this PR with the sed replacements increases the complexity even further. So, ideally I would like to make some commitment to refactor it in a follow up, or do a preparatory refactoring PR.

Reply:
@astefanutti Could you open dedicated issue?

Reply:
Sure, I've just created #1798 to capture it.